### PR TITLE
feat: Ability to Share Service Instances

### LINKED ISSUE #437: [FEATURE] Ability to Share Service Instances
### What area do you want to see improved?

terraform provider

### Is your feature request related to a problem? Please describe.

I'm moving from a CloudFoundry environment into a Kyma environment, and it would be very helpful to enable a service to be consumed by both Kyma (via BTP Service Operator) and CloudFoundry at the same time.

My understanding is that this can be done via the [btp share services/instance command on the BTP CLI](https://help.sap.com/docs/btp/btp-cli-command-reference/btp-share-services-instance). However, I can't find any way to do this with the existing BTP TF provider. One can read whether the service is shared, but cannot set whether it is shared. As a result, I have to share the service via the BTP CLI as a separate command in order to then pick up the service in my Kyma environment Terraform.

### Describe the solution you would like

On the [`subaccount_service_instance` resource](https://registry.terraform.io/providers/SAP/btp/latest/docs/resources/subaccount_service_instance), I think it would make sense to have a value, `shared`, that defaults to `false`. When set to `true`, is reconciled by performing this command.

### Describe alternatives you have considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by diya-dhan**
## Purpose

Users now have the flexibility to share their service instances across different environments by modifying the `shared` parameter within the `Service Instance` resource schema. This parameter can be adjusted during the instance creation or updated later as per the user's preference.

During both the creation and updating processes of the service instance resource, an extra call is made to the 'btp share' and 'btp unshare' APIs based on the value specified for the 'shared' parameter.

It's crucial for users to ensure that the instances they intend to share are associated with a service plan that supports sharing.

fix #437 

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type

What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test

* Test the code via automated test

```bash
go test ./...
```

## What to Check

Verify that the following are valid:

* Automated tests are executed successfully

## Checklist for reviewer

The following organizational tasks must be completed before merging this PR:

* [x] The PR is assigned to the Terraform project and a status is set (typically "in review").
* [x] The PR has the matching labels assigned to it.
* [x] The PR has a milestone assigned to it.
* [x] If the PR closes an issue, the issue is referenced.
* [x] Possible follow-up items are created and linked.

### CODE REVIEW

File: docs/resources/subaccount_service_instance.md [hunk 1]
```diff
@@ -45,8 +45,16 @@ resource "btp_subaccount_service_instance" "alert_notification_free" {
...
 
+# create an instance of the destination service and also share the instance
+resource "btp_subaccount_service_instance" "destination_lite" {
+  subaccount_id = "6aa64c2f-38c1-49a9-b2e8-cf9fea769b7f"
+  # The service plan ID can be looked up via the data source btp_subaccount_service_plan
+  serviceplan_id = "89ae3076-9c50-4c86-a917-7953c2e75cf2" # destination - lite
+  name           = "destination-instance-new"
+  shared         = true
+}
+```
```
Comment (lechnerc77) [inline docs/resources/subaccount_service_instance.md]: This part needs to be put into `examples/resources/btp_subaccount_service_instance/resource.tf` as additional example.
Then execute `make generate` to re-generate the documentation

File: internal/provider/resource_subaccount_service_instance.go [hunk 2]
```diff
@@ -78,6 +79,11 @@ func (rs *subaccountServiceInstanceResource) Schema(ctx context.Context, _ resou
 					jsonvalidator.ValidJSON(),
 				},
 			},
+			"shared": schema.BoolAttribute{
+				MarkdownDescription: "Shows whether the service instance is shared. This parameter can also be used to allow the instance to be shared between different environments. Ensure that the instance is created with a plan that supports instance sharing.",
```
Comment (lechnerc77) [inline internal/provider/resource_subaccount_service_instance.go]: I would propose a different text here: "The configuration parameter for service instance sharing. Ensure that the instance is created with a plan that supports instance sharing."

**Review by lechnerc77 — CHANGES_REQUESTED** [review]
Two minor changes concerning documentation and providing the example

File: internal/btpcli/client.go [hunk 3]
```diff
@@ -310,12 +310,13 @@ func (v2 *v2Client) Execute(ctx context.Context, cmdReq *CommandRequest, options
...
 	if cmdRes.StatusCode >= 400 {
+
 		var backendError struct {
-			Message string `json:"error"`
+			Message     string `json:"error"`
 		}
 
 		if err = json.NewDecoder(res.Body).Decode(&backendError); err == nil {
-			err = fmt.Errorf(backendError.Message)
+			err = fmt.Errorf("error: %s", backendError.Message)
```
Comment (lechnerc77) [inline internal/btpcli/client.go]: As you changed this error message logging you will probably also need to adjust some test cases with respect to the expected error message format

File: docs/resources/subaccount_service_instance.md (hunk 1 above)
Comment (diya-dhan) [reply docs/resources/subaccount_service_instance.md]: change has been made

File: internal/provider/resource_subaccount_service_instance.go (hunk 2 above)
Comment (diya-dhan) [reply internal/provider/resource_subaccount_service_instance.go]: change has been made

File: internal/btpcli/client.go (hunk 3 above)
Comment (diya-dhan) [reply internal/btpcli/client.go]: I've reverted this change for the time being

**Comment by lechnerc77** [comment]
Inclusive Language Check failure is accepted as errors occur due to service responses we cannot influence

**Comment by lechnerc77** [comment]
Sonar Cloud findings are acceptable

**Review by lechnerc77 — APPROVED** [review]
lgtm

**Pull request merged**